### PR TITLE
[SQL] Throw an error for correlated exists/IN subqueries with LIMIT clauses

### PR DESCRIPTION
**Pull request by agubichev**
### What changes were proposed in this pull request?

Support for LIMIT in correlated lateral and scalar subqueries was added in https://github.com/apache/spark/pull/42705. However, that PR also extended Analyzer support for limit/order by in EXISTS/IN subqueries, for which we don't have good planning support yet. 
This PR introduces checks that ensure correlated IN/EXISTS subqueries do not have LIMIT clauses.
We do not add a new type of QueryCompilationError, as we plan to fully support such subqueries in the next couple of months, so for now we re-use the same analysis error as used to be issued before https://github.com/apache/spark/pull/42705.

### Why are the changes needed?

See above.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Unit tests

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Review by jchen5 — APPROVED** [review]
@allisonwang-db @cloud-fan

**Comment by jchen5** [comment]
Note that after these two PRs, the state is as follows:
- Newly enabled: Correlated scalar/lateral subqueries with LIMIT, which required new logic in DecorrelateInnerQuery
- Newly enabled: Uncorrelated IN/EXISTS subqueries with LIMIT, which did not require any new logic other than allowing them through CheckAnalysis.
- Still disabled: Correlated IN/EXISTS subqueries with LIMIT.

**Comment by agubichev** [comment]
> * Uncorrelated IN/EXISTS subqueries with LIMIT

Yes, these are all covered by tests. The first class of queries are (correlated scalar/lateral with limit) are tested in https://github.com/apache/spark/pull/42705.
The second class of queries (correlated exists/in with limit) are tested in this PR.
The third class of queries (uncorrelated exists/in with limit) are tested at HEAD, the tests in this PR are largely based on those tests.

**Comment by cloud-fan** [comment]
Can we create a JIRA ticket for it and make all tests pass? Then it's ready to go

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala [hunk 1]
```diff
@@ -283,6 +284,14 @@ object PullupCorrelatedPredicates extends Rule[LogicalPlan] with PredicateHelper
         } else {
           a
         }
+      case l @ Limit(_, _) if predicateMap.nonEmpty =>
```
[the remark this answers is not in the document]
Comment (jchen5) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:287]: IIRC in CheckAnalysis we don't easily know whether it's inside IN/EXISTS (unsupported) or scalar/lateral (supported).

**Comment by agubichev** [comment]
Closing this as all the cases are being handled by https://github.com/apache/spark/pull/43111

**Comment by jchen5** [comment]
Hmm, it's true they will be handled but might still be good to have this check in case a user disables the flag for new support?

**Comment by agubichev** [comment]
> Hmm, it's true they will be handled but might still be good to have this check in case a user disables the flag DECORRELATE_EXISTS_AND_IN_SUBQUERIES to force the old codepath. So that we throw an exception (as we did before) instead of giving wrong results.

I'll just add it there directly!

**Comment by jchen5** [comment]
Great, sounds good

**Pull request closed without merging**